cooked-validators/tests/Cooked/Tx/Constraints/TypeSpec.hs

```diff
@@ -0,0 +1,275 @@
+    oneof
+      [ TxLabel @String <$> arbitrary,
+        TxLabel @Integer <$> arbitrary
+      ]
+
+instance Arbitrary TxOpts where
+  arbitrary =
+    TxOpts <$> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+
+-- ** Generating minting policies
+
+-- At the moment, there is only one policy we generate:
+
+{-# INLINEABLE mkMockPolicy #-}
+mkMockPolicy :: Pl.BuiltinData -> Pl.BuiltinData -> ()
+mkMockPolicy _ _ = ()
+
+mockPolicy :: Pl.MintingPolicy
+mockPolicy = Pl.mkMintingPolicyScript $$(Pl.compile [||mkMockPolicy||])
+
+instance Arbitrary Pl.MintingPolicy where
+  arbitrary = return mockPolicy
+
+-- ** Generating 'MintsConstraint's
+
+instance Arbitrary Pl.TokenName where
+  arbitrary = hedgehog LG.genTokenName
+
+instance Arbitrary MintsConstraint where
+  arbitrary =
+    MintsConstraint @Bool <$> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+
+-- ** Generating 'Value's
+
+instance Arbitrary Pl.Value where
+  arbitrary = hedgehog LG.genValue
+
+-- ** Generating 'Address'es
+
+instance Arbitrary Pl.PubKeyHash where
+  arbitrary = Pl.PubKeyHash <$> Pl.BuiltinByteString <$> arbitrary
+
+instance Arbitrary Pl.StakePubKeyHash where
+  arbitrary = Pl.StakePubKeyHash <$> arbitrary
+
+genPKCredential :: Gen Pl.Credential
+genPKCredential = Pl.PubKeyCredential <$> arbitrary
+
+instance Arbitrary Pl.ValidatorHash where
+  arbitrary = Pl.ValidatorHash <$> Pl.BuiltinByteString <$> arbitrary
+
+genScriptCredential :: Gen Pl.Credential
+genScriptCredential = Pl.ScriptCredential <$> arbitrary
+
+instance Arbitrary Pl.Credential where
+  arbitrary = oneof [genPKCredential, genScriptCredential]
+
+instance Arbitrary Pl.StakingCredential where
+  arbitrary =
+    oneof
+      [ Pl.StakingHash <$> arbitrary,
+        Pl.StakingPtr <$> arbitrary <*> arbitrary <*> arbitrary
+      ]
+
+-- TODO: Do we have to impose some restrictions on the staking credentials in
+-- 'genPKAddress' and 'genScriptAddress'?
+
+genPKAddress :: Gen Pl.Address
+genPKAddress = Pl.Address <$> genPKCredential <*> arbitrary
+
+genScriptAddress :: Gen Pl.Address
+genScriptAddress = Pl.Address <$> genScriptCredential <*> arbitrary
+
+instance Arbitrary Pl.Address where
+  arbitrary = oneof [genPKAddress, genScriptAddress]
+
+-- ** Generating (typed) validators
+
+type ADatum = Bool
+
+type ARedeemer = Bool
+
+data AContract
+
+instance Pl.ValidatorTypes AContract where
+  type DatumType AContract = ADatum
+  type RedeemerType AContract = ARedeemer
+
+{-# INLINEABLE mkAValidator #-}
+mkAValidator :: ADatum -> ARedeemer -> Pl.ScriptContext -> Bool
+mkAValidator _ _ _ = True
+
+aValidator :: Pl.TypedValidator AContract
+aValidator =
+  Pl.mkTypedValidator @AContract
+    $$(Pl.compile [||mkAValidator||])
+    $$(Pl.compile [||wrap||])
+  where
+    wrap = Pl.mkUntypedValidator @ADatum @ARedeemer
+
+instance Arbitrary (Pl.TypedValidator AContract) where
```

Comment:
I am not sure about what kind of arbitrary instance we would like to have for typed validators.

If I understand correctly, for now we can only generate the `aValidator` which always accepts transactions but we would ideally like to be able to generate any arbitrary validator given for example a datum type and redeemer type? Do we want also to generate arbitrary datum and redeemer types?
Or, do we want to restrict ourselves to some catalog of validators?